### PR TITLE
Allow to prewarm mark cache for data parts

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a setting `prewarm_mark_cache` which enables loading of marks to mark cache on inserts, merges, fetches of parts and on startup of the table.

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 32cfa6de6a83d3c4ed6b18f4e16310e0fb0ef4c6 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71053/32cfa6de6a83d3c4ed6b18f4e16310e0fb0ef4c6/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/71053/32cfa6de6a83d3c4ed6b18f4e16310e0fb0ef4c6/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -2335,6 +2336,54 @@ void MergeTreeData::stopOutdatedAndUnexpectedDataPartsLoadingTask()
...
+    std::sort(data_parts.begin(), data_parts.end(), [&to_tuple](const auto & lhs, const auto & rhs)
+    {
+        return to_tuple(lhs) > to_tuple(rhs);
+    });
+
+    ThreadPoolCallbackRunnerLocal<void> runner(pool, "PrewarmMarks");
+
+    for (const auto & part : data_parts)
+    {
+        if (mark_cache->sizeInBytes() >= mark_cache->maxSizeInBytes() * 0.95)
```
Comment (kssenii) [inline src/Storages/MergeTree/MergeTreeData.cpp:2376]: should a setting be added for this ratio to be changed if needed?

Comment (CurtizJ) [reply src/Storages/MergeTree/MergeTreeData.cpp:2376]: Yes, we can add a setting.

**Review by kssenii — APPROVED** [review]

File: src/Interpreters/InterpreterSystemQuery.cpp [hunk 2]
```diff
@@ -1298,6 +1307,28 @@ RefreshTaskList InterpreterSystemQuery::getRefreshTasks()
...
+    if (table_id.empty())
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Table is not specified for prewarming marks cache");
+
+    auto table_ptr = DatabaseCatalog::instance().getTable(table_id, getContext());
+    auto * merge_tree = dynamic_cast<MergeTreeData *>(table_ptr.get());
+
+    if (!merge_tree)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Command PREWARM MARK CACHE is supported only for MergeTree table, but got: {}", table_ptr->getName());
+
+    ThreadPool pool(
```
[the remark this answers is not in the document]
Comment (CurtizJ) [reply src/Interpreters/InterpreterSystemQuery.cpp:1323]: I think local thread pool is ok here. This command is not supposed to be executed frequently and concurrently. Also it resembles the logic of regular pipeline executor which have local thread pool with `max_threads`.

**Pull request merged**